next/src/validation/schema.ts

```
@@ -0,0 +1,106 @@
+ * @param schema - The schema to validate against
+ * @returns An array of validation errors
+ * @description
+ * - If the schema type is an array, the value must be an instance of one of the types in the array.
+ * - If the schema type is a string, the value must be of the same type.
+ */
+function validateType(value: SchemaValue, schema: JsfSchema): ValidationError[] {
+  const schemaType = getSchemaType(schema)
+  const valueType = value === undefined ? 'undefined' : typeof value
+
+  if (Array.isArray(schemaType) ? !schemaType.includes(valueType) : valueType !== schemaType) {
+    return [{ path: [], validation: 'type', message: `should be ${schemaType}` }]
+  }
+
+  return []
+}
+
+/**
+ * Validate a value against a schema
+ * @param value - The value to validate
+ * @param schema - The schema to validate against
+ * @param required - Whether the value is required
+ * @returns An array of validation errors
+ * @description
+ * This function is the main validation function to validate a value against a schema.
+ * - It validates boolean schemas
+ * - It validates the `required` constraint
+ * - It validates the type of the value
+ * - It delegates to type specific validation functions such as `validateObject` and `validateString`
+ * @see `validateType` for type validation
+ */
+export function validateSchema(value: SchemaValue, schema: JsfSchema, required: boolean = false): ValidationError[] {
+  if (value === undefined && required) {
+    return [{ path: [], validation: 'required', message: 'is required' }]
+  }
+
+  if (value === undefined) {
+    return []
+  }
+
+  if (typeof schema === 'boolean') {
+    return schema ? [] : [{ path: [], validation: 'valid', message: 'always fails' }]
+  }
+
+  const typeValidationErrors = validateType(value, schema)
+  if (typeValidationErrors.length > 0) {
+    return typeValidationErrors
+  }
+
+  const errors = [
+    ...validateObject(value, schema),
+    ...validateString(value, schema),
+  ]
```

Assuming this is where we would list all our new validation functions. 

One thing I wanted to ask is would we keep using yup for validation or do we go vanilla for everything?

---

> One thing I wanted to ask is would we keep using yup for validation or do we go vanilla for everything?
I think we should ditch yup, IMO it does not really bring much value.

---

@thien-remote 
> Is there a case where we would expect errors from multiple validation here?

Good point. This is just a stylistic choice. I think there two ways to do this:
1. Have one if - else if block in `validateSchema` that checks what type the schema specifies and call the corresponding validate function
2. Call all validate functions where they do that check and will either validate or not

I'm leaning towards the seconds because I'm thinking there might be cases where the condition whether to run the validations is more complex. Not sure though tbh, it's just a hunch. I don't feel strongly about this.

---

> Just one thing we need to keep in mind is that currently the yupSchema is responsible for the "nice errors" that we get after validation. We'll need to keep those somehow.

> I suppose the plan is to actually implement them? Sorry Luka but this I need you 😆

Ah sorry, forgot to hit send on a comment I had already written 😅
Yes, we need to return the same errors indeed. I'll follow up on this MR with building fields for the schema. That's where would also like to address this.

---

Took me a good second to understand what you mean here Luka but I think I got it 😄 

>1. Have one if - else if block in validateSchema that checks what type the schema specifies and call the corresponding validate function
>2. Call all validate functions where they do that check and will either validate or not
>
>I'm leaning towards the seconds because I'm thinking there might be cases where the condition whether to run the validations is more complex. Not sure though tbh, it's just a hunch. I don't feel strongly about this.

I **think** also like the second approach here since it would keep this function clean (less if statements) and we could always just check in the individual validation function whether they need to run or not. Might be bit repetitive overall but it is still clearer overall IMO.

One thing that would change my opinion here is if that condition can change on the long term and become more complex in which case we would potentially need to update all the validation functions. You see where I'm going with this..

---

I suggest we keep it like this then.

> One thing that would change my opinion here is if that condition can change on the long term and become more complex

I don't think this will be the case. But if we run into that problem, it won't be too bad. We won't have hundreds of validation functions.